File: fn/src/main/java/org/sqlbuilder/fn/objects/Governor.java
package org.sqlbuilder.fn.objects;

import org.sqlbuilder.annotations.RequiresIdFrom;
import org.sqlbuilder.common.*;

import java.util.Comparator;

import edu.berkeley.icsi.framenet.GovernorType;

public class Governor implements HasId, Insertable
{
	public static final Comparator<Governor> COMPARATOR = Comparator.comparing(Governor::getWord).thenComparing(Governor::getType);

	public static final SetCollector2<Governor> COLLECTOR = new SetCollector2<>(COMPARATOR);

	private final String type;

	private final Word word;

	public static Governor make(final GovernorType governor)
	{
		var g = new Governor(governor);
		COLLECTOR.add(g);
		return g;
	}

	private Governor(final GovernorType governor)
	{
		this.type = governor.getType();
		this.word = Word.make(governor.getLemma());
	}

	// A C C E S S

	public String getType()
	{
		return type;
	}

	public String getWord()
	{
		return word.getWord();
	}

	@RequiresIdFrom(type = Governor.class)
	@Override
	public Integer getIntId()
	{
		return COLLECTOR.apply(this);
	}

	// I N S E R T

	@RequiresIdFrom(type = Word.class)
	@Override
	public String dataRow()
	{
		// governorid,governortype,fnwordid
		return String.format("'%s',%s", //
				Utils.escape(type), //
				word.getSqlId());
	}

	@Override
	public String comment()
	{
		return String.format("word=%s", word.getWord());
	}

	@Override
	public String toString()
	{
		return String.format("[GOV type=%s word=%s]", type, word);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/objects/Governor.kt
package org.sqlbuilder.fn.objects

import edu.berkeley.icsi.framenet.GovernorType
import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.HasId
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.SetCollector
import org.sqlbuilder.common.Utils.escape
import org.sqlbuilder.fn.objects.Word.Companion.make

class Governor private constructor(
    governor: GovernorType,
) : HasId, Insertable {

    val type: String = governor.getType()

    private val word: Word = make(governor.getLemma())

    fun getWord(): String {
        return word.word
    }

    @RequiresIdFrom(type = Governor::class)
    override fun getIntId(): Int {
        return COLLECTOR.invoke(this)
    }

    // I N S E R T

    @RequiresIdFrom(type = Word::class)
    override fun dataRow(): String {
        return "'${escape(type)}',${word.getSqlId()}"
    }

    override fun comment(): String {
        return "word=${word.word}"
    }

    override fun toString(): String {
        return "[GOV type=$type word=${word}]"
    }

    companion object {

        @JvmField
        val COMPARATOR: Comparator<Governor> = Comparator
            .comparing<Governor, String> { it.getWord() }
            .thenComparing<String> { it.type }

        @JvmField
        val COLLECTOR = SetCollector<Governor>(COMPARATOR)

        fun make(governor: GovernorType): Governor {
            val g = Governor(governor)
            COLLECTOR.add(g)
            return g
        }
    }
}